### ISSUE TITLE
TypeVar polluted by `T | None` return/default annotation into key callable param

### ISSUE THREAD

**Issue by jvdillon** [astral-sh/ty#4016]
### Summary

```python
from typing import Callable, TypeVar
T = TypeVar("T")

def pick(xs: list[T], key: Callable[[T], object], default: T | None = None) -> T | None:
    return default

def u(c: list[str]) -> str | None:
    return pick(c, key=lambda p: p.upper())   # p wrongly inferred str | None


# warning[unresolved-attribute]: Attribute `upper` is not defined on `None` in union `str | None`
#  --> /home/jvdillon/ty_issue.py:8:34
#   |
# 8 |     return pick(c, key=lambda p: p.upper())   # p wrongly inferred str | None
#   |                                  ^^^^^^^
#   |
#
# Found 1 diagnostic

```

But this is wrong -- `p` cannot be `None` here. `pick`'s param is `key: Callable[[T], object]`. `T` is bound by `xs: list[str]` --> `T = str` so key must be `Callable[[str], object]` and `p is str` never `None`.

Closest issues are maybe #3934 and #2056 but those seem different?

### Version

0.0.60

**Comment by carljm** [comment]
Thanks for the report! This is a valid issue, I think the solution is actually described by #3933. Will keep this open as a concrete sub-issue.

**Comment by carljm** [comment]
Moving this to stable, since it's cropped up several times in short order following the release, and also showed up in internal codebases as a notable regression.

**Comment by ibraheemdev** [comment]
The root cause here is [this TODO in the call binding logic](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/src/types/call/bind.rs#L6476). We currently do not store the declared type constraints as a constraint-set, and so aren't able to correctly conjoin it with the argument constraints, instead relying on a naive preference of the outer declared type with a fallback to the argument constraints. That is blocked on https://github.com/astral-sh/ty/issues/3559, I don't think this needs the full design described by https://github.com/astral-sh/ty/issues/3933.

**Comment by sharkdp** [comment]
I see this happening with `sorted` a lot, so I'm mentioning this here for better discoverability.

Here's a case that I think is related to this issue as well:
```py
from typing import Iterable

items: list[str] = ["b", "A"]

# Attribute `lower` is not defined on `None` in union `str | None`
items_sorted: Iterable[str | None] = sorted(items, key=lambda item: item.lower())
```

**Comment by Yay295** [comment]
I ran into this with discord.py's `discord.utils.find()`:

```python
discord.utils.find(lambda m: unorm('NFKD',m.name.casefold()) == unorm('NFKD',value.casefold()), guild.members)
```

which, without discord.py classes, is essentially

```python
from typing import Any, Callable, Iterable, Optional, TypeVar

T = TypeVar('T')
def find(predicate: Callable[[T], Any], iterable: Iterable[T]) -> Optional[T]:
    return next(filter(predicate, iterable), None)

def test() -> Optional[str]:
    data = ['1','2','3']
    result = find(lambda x: x.casefold() == '2', data) # no error
    if result: return result
    return find(lambda x: x.casefold() == '2', data) # Attribute `casefold` is not defined on `None` in union `str | None`
```
